File: typescript-workshop/part-4-behavior/tests/helper/InMemoryOrderRepository.ts
import { Order } from '../../src/domain/Order';
import { OrderStatus } from '../../src/domain/OrderStatus';
import { OrderRepository } from '../../src/repository/OrderRepository';

export class InMemoryOrderRepository implements OrderRepository {
  private readonly orders = new Map<string, Order>();

  findById(id: string): Order | null {
    return this.orders.get(id) ?? null;
  }

  save(order: Order): void {
    this.orders.set(order.id, order);
  }

  updateStatus(id: string, status: OrderStatus): void {
    const order = this.orders.get(id);
    if (order !== undefined) {
      this.orders.set(id, new Order(
        order.id,
        order.customerId,
        order.items,
        status,
        order.totalPrice,
        order.transactionId,
        order.trackingNumber,
      ));
    }
  }

  with(order: Order): this {
    this.orders.set(order.id, order);
    return this;
  }
}

File: csharp-workshop/Part4.Behavior.Tests/Helper/InMemoryOrderRepository.cs
using Part4.Behavior.Domain;
using Part4.Behavior.Repository;

namespace Part4.Behavior.Tests.Helper;

public class InMemoryOrderRepository : IOrderRepository
{
    private readonly Dictionary<string, Order> _orders = new();

    public InMemoryOrderRepository With(Order order)
    {
        _orders[order.Id] = order;
        return this;
    }

    public Order? FindById(string orderId)
    {
        _orders.TryGetValue(orderId, out var order);
        return order;
    }

    public void UpdateStatus(string orderId, OrderStatus status)
    {
        if (_orders.TryGetValue(orderId, out var order))
        {
            _orders[orderId] = new Order(
                order.Id, order.CustomerId, order.Items,
                status, order.TotalPrice, order.TransactionId, order.TrackingNumber
            );
        }
    }

    public void Save(Order order)
    {
        _orders[order.Id] = order;
    }
}
